server/websocket.go

```diff
@@ -0,0 +1,845 @@
+// Copyright 2020 The NATS Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"bytes"
+	"compress/flate"
+	"crypto/sha1"
+	"crypto/tls"
+	"encoding/base64"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+	"unicode/utf8"
+)
+
+type wsOpCode int
+
+const (
+	// From https://tools.ietf.org/html/rfc6455#section-5.2
+	wsTextMessage   = wsOpCode(1)
+	wsBinaryMessage = wsOpCode(2)
+	wsCloseMessage  = wsOpCode(8)
+	wsPingMessage   = wsOpCode(9)
```

Comment:
btw is there a default ping interval timer from the server to the clients? Since websockets via https could be used with load balancers, maybe the default read timeout from an LB (60s in aws by default for example) could hit earlier than the ping/pong interval from NATS (default 2 min).

Reply:
In current implementation, I am not sending WS pings since I relied on the fact that server would send NATS PINGs. The server will respond to client PINGs if sent, but it won't initiate WS pings. I am opened to implement that, or another approach would be to document that users planning to use websocket should lower the ping interval?

Reply:
I think that sounds good, maybe first leave as is and just document that ping interval ought to be set lower on the client/lb for now before baking that into the server.